application/src/main/java/org/thingsboard/server/controller/FirmwareController.java

```diff
@@ -129,18 +129,17 @@ public Firmware saveFirmwareData(@PathVariable(FIRMWARE_ID) String strFirmwareId
             firmware.setVersion(info.getVersion());
             firmware.setAdditionalInfo(info.getAdditionalInfo());
 
-            byte[] data = file.getBytes();
-            if (StringUtils.isEmpty(checksumAlgorithm)) {
-                checksumAlgorithm = "sha256";
-                checksum = Hashing.sha256().hashBytes(data).toString();
+            ByteBuffer data = ByteBuffer.wrap(file.getBytes());
+            if (StringUtils.isEmpty(checksum)) {
+                checksum = firmwareService.generateChecksum(checksumAlgorithm, data);
             }
 
             firmware.setChecksumAlgorithm(checksumAlgorithm);
             firmware.setChecksum(checksum);
             firmware.setFileName(file.getOriginalFilename());
             firmware.setContentType(file.getContentType());
-            firmware.setData(ByteBuffer.wrap(data));
-            firmware.setDataSize((long) data.length);
+            firmware.setData(data);
+            firmware.setDataSize((long) data.array().length);
```

Comment:
This is wrong. Also it will work in this case, it may stop working in the future, if someone will manipulate with the ByteBuffer. The .array() is internal data structure.